python/lsst/ip/isr/isrTask.py

```diff
@@ -918,22 +913,31 @@ def readIsrData(self, dataRef, rawExposure):
                         if self.config.doFlat else None)
 
         brighterFatterKernel = None
+        brighterFatterGains = None
         if self.config.doBrighterFatter is True:
-
-            # Use the new-style cp_pipe version of the kernel is it exists.
             try:
+                # Use the new-style cp_pipe version of the kernel is it exists
+                # If using a new-style kernel, always use the self-consistent
+                # gains, i.e. the ones inside the kernel object itself
                 brighterFatterKernel = dataRef.get("brighterFatterKernel")
+                brighterFatterGains = brighterFatterKernel.gain
+                self.log.info("New style bright-fatter kernel (brighterFatterKernel) loaded")
             except NoResults:
-                # Fall back to the old-style numpy-ndarray style kernel if necessary.
-                try:
+                try:  # Fall back to the old-style numpy-ndarray style kernel if necessary.
                     brighterFatterKernel = dataRef.get("bfKernel")
+                    self.log.info("Old style bright-fatter kernel (np.array) loaded")
                 except NoResults:
                     brighterFatterKernel = None
             if brighterFatterKernel is not None and not isinstance(brighterFatterKernel, numpy.ndarray):
-                # If the kernel is not an ndarray, it's the cp_pipe version, so extract the kernel for
-                # this detector, or raise an error.
+                # If the kernel is not an ndarray, it's the cp_pipe version
+                # so extract the kernel for this detector, or raise an error
                 if self.config.brighterFatterLevel == 'DETECTOR':
-                    brighterFatterKernel = brighterFatterKernel.kernel[ccd.getId()]
+                    if brighterFatterKernel.detectorKernel:
+                        brighterFatterKernel = brighterFatterKernel.detectorKernel[ccd.getId()]
+                    elif brighterFatterKernel.detectorKernelFromAmpKernels:
+                        brighterFatterKernel = brighterFatterKernel.detectorKernelFromAmpKernels[ccd.getId()]
+                    else:
+                        raise RuntimeError("Failed to extract kernel from new-style BF kernel.")
                 else:
                     # TODO DM-15631 for implementing this
                     raise NotImplementedError("Per-amplifier brighter-fatter correction not implemented")
```

Comment:
Can you duplicate this logic into the gen3 runQuantum() method (line isrTask.py:827)?  The kernel object should be supplied as inputs['bfKernel'], and can be overwritten in place by the appropriate detectorKernel (the linearizer gets and uses a detector as well).  Adding the inputs['bfGains'] entry should ensure this works cleanly in that case.

Reply:
I was about to try, but I have no way of testing what I've done as I don't have any Gen3 repos (or even knowledge yet), and combined with no experience in such things, I think it's best left. I'll ticket, leave a `TODO:` and assign it to you. Sorry and thanks!

Reply:
DM-22776